src/main/java/org/jenkins/plugins/lockableresources/LockStep.java

```diff
@@ -0,0 +1,65 @@
+package org.jenkins.plugins.lockableresources;
+
+import java.io.Serializable;
+
+import org.jenkinsci.plugins.workflow.steps.AbstractStepDescriptorImpl;
+import org.jenkinsci.plugins.workflow.steps.AbstractStepImpl;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+
+import hudson.Extension;
+import hudson.model.AutoCompletionCandidates;
+
+public class LockStep extends AbstractStepImpl implements Serializable {
+
+	public final String resource;
+
+	public boolean inversePrecedence = false;
+
+	@DataBoundConstructor
+	public LockStep(String resource) {
+		if (resource == null || resource.isEmpty()) {
+			throw new IllegalArgumentException("must specify resource");
+		}
+		if (LockableResourcesManager.get().fromName(resource) == null) {
+			throw new IllegalArgumentException("resource [" + resource + "] does not exist, missing global definition?");
```

Comment:
> unless we somehow uniquely identify a `lock` step

Huh? The step is already specifying a resource name. My suggestion was merely to create such a resource if it does not already exist.

```java
LockableResourcesManager mgr = LockableResourcesManager.get();
if (mgr.fromName(resource) == null) {
    mgr.getResources().add(new LockableResource(resource));
    mgr.save();
}
```

Something like that.

> Freestyle interoperability would require to create resources in global configuration.

Sure, but for a Pipeline-only installation, you would not need this step. I would not expect most users of this plugin to be mixing freestyle and Pipeline builds on the same resource; it is probably a corner case.

> This case is similar to `tool`

Not really. In the case of a `ToolInstallation`, there is real configuration you are providing, like a `ToolInstaller`. The name of the tool is merely an identifier for that configuration. A resource in the common case is just an opaque token. `LockableResource` can apparently take a description and some labels, but these seem quite optional.